tests/test_agent.py

```diff
@@ -0,0 +1,247 @@
+    dev_id = agent.generate_device_id()
+    assert dev_id
+
+
+def test_get_device_id(cert):
+    with mock.patch(
+            'builtins.open',
+            mock.mock_open(read_data=cert),
+            create=True
+    ):
+        device_id = agent.get_device_id()
+        assert device_id == 'localhost'
+
+
+def test_generate_cert():  # TODO: parse key and csr
+    cert = agent.generate_cert('localhost')
+    assert cert['key']
+    assert cert['csr']
+
+
+@pytest.mark.vcr
+def test_get_ca_cert():
+    ca_bundle = agent.get_ca_cert()
+    assert "BEGIN CERTIFICATE" in ca_bundle
+
+
+def test_get_ca_cert_none_on_fail():
+    with mock.patch('requests.get') as req,\
+            mock.patch('builtins.print') as prn:
+        req.return_value.ok = False
+        ca_bundle = agent.get_ca_cert()
+    assert ca_bundle is None
+    assert mock.call('Failed to get CA...') in prn.mock_calls
+    assert prn.call_count == 3
+
+
+def test_get_open_ports(nmap_fixture, netif_gateways, netif_ifaddresses):
+    with mock.patch('agent.security_helper.nmap') as nmap, \
+            mock.patch('netifaces.gateways') as gw, \
+            mock.patch('netifaces.ifaddresses') as ifaddr:
+        nmap.return_value.stdout = nmap_fixture
+        gw.return_value = netif_gateways
+        ifaddr.return_value = netif_ifaddresses
+        result = agent.get_open_ports()
+        assert len(result) == 4  # TODO: test keys/values
+
+
+@pytest.mark.vcr
+def test_send_ping(raspberry_cpuinfo, uptime, tmpdir, cert, key):
+    crt_path = tmpdir / 'client.crt'
+    key_path = tmpdir / 'client.key'
+    Path(crt_path).write_text(cert)
+    Path(key_path).write_text(key)
+    agent.CERT_PATH = str(tmpdir)
+    agent.CLIENT_CERT_PATH = str(crt_path)
+    agent.CLIENT_KEY_PATH = str(key_path)
+    with mock.patch(
+            'builtins.open',
+            mock.mock_open(read_data=raspberry_cpuinfo),
+            create=True
+    ), \
+        mock.patch('socket.getfqdn') as getfqdn,\
+            mock.patch('builtins.print') as prn, \
+            mock.patch(
+                'builtins.open',
+                mock.mock_open(read_data=uptime),
+                create=True
+            ):
+        getfqdn.return_value = 'localhost'
+        ping = agent.send_ping()
+        assert ping is None
+        assert prn.call_count == 0 or (prn.call_count == 1 and mock.call('Ping failed.') in prn.mock_calls)
+
+
+@pytest.mark.vcr
+def test_say_hello(tmpdir, cert, key):
+    crt_path = tmpdir / 'client.crt'
+    key_path = tmpdir / 'client.key'
+    Path(crt_path).write_text(cert)
+    Path(key_path).write_text(key)
+    agent.CERT_PATH = str(tmpdir)
+    agent.CLIENT_CERT_PATH = str(crt_path)
+    agent.CLIENT_KEY_PATH = str(key_path)
+    with mock.patch('builtins.print') as prn:
+        with pytest.raises(json.decoder.JSONDecodeError):
+            hello = agent.say_hello()
+        assert mock.call('Hello failed.') in prn.mock_calls
+
+
+def test_uptime(uptime):
```

Comment:
I don't fully understand this one. Does uptime always end up being 60s when you mock the interface?

Reply:
It's pre-populated value.
Please check https://github.com/slilo/agent/blob/tests/tests/conftest.py#L160
It passed in the arguments (pytest magic).